sample/weighted.go

```diff
@@ -0,0 +1,98 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import "math/rand"
+
+// Weighted provides sampling without replacement from a collection of items with
+// weighted probabilites.
+type Weighted struct {
+	weights []float64
+	heap    []float64
+	src     *rand.Rand
+}
+
+// NewWeighted returns a Weighted for the weights w. Elements of w are modified
+// during sampling, so a copy should be passed if the user needs to keep the values.
+// If src is nil, rand.Rand is used as the random source.
+func NewWeighted(w []float64, src *rand.Rand) Weighted {
```

Comment:
I think NewWeighted should allocate and copy instead of maintaining a reference. Everything is much easier to think about when data isn't shared, and the ReweightAll method allows the Weighted to be continually reused to limit garbage.